### DISCUSSION TITLE
Purpose of bidirectional rewrite!

### DISCUSSION THREAD

**Discussion by silversquirl** [egraphs-good/egg#222] [Q&A] [answered]
`rewrite!` has a bidirectional variant: `rewrite!("name"; "pattern" <=> "subst")`

However, egg will happily apply a regular unidirectional rule in reverse. What is the purpose of the bidirectional form of `rewrite!`?

**Comment by mwillsey** [comment, marked as answer]
egg will not apply a unidirectional rewrite in both directions. However,
the equivalences found by a rewrite (uni- or bi-directional) are
bidirectional. For example, if your e-graph contains only `x + x` and you
run the rewrite `2 * x => x + x`, nothing will happen. However if both `x +
x` and `2 * x` were in the e-graph to begin with, then it would fire.
